# dupsifter

`dupsifter` is a command line tool for marking PCR duplicates in both WGS and WGBS
datasets. It is based on the [samblaster](https://github.com/GregoryFaust/samblaster)
methodology for finding and marking duplicates.

## Publication

If you use dupsifter, kindly [cite](https://doi.org/10.1093/bioinformatics/btad729):

```
Jacob Morrison, Wanding Zhou, Benjamin K Johnson, Hui Shen,
Dupsifter: a lightweight duplicate marking tool for whole genome bisulfite sequencing,
Bioinformatics, Volume 39, Issue 12, December 2023, btad729,
https://doi.org/10.1093/bioinformatics/btad729
```

## Documentation

Documentation, including installation and usage, can be found in the `docs/` directory and at
[https://huishenlab.github.io/dupsifter/](https://huishenlab.github.io/dupsifter/).

## Download and Install

All releases are available on [GitHub](https://github.com/huishenlab/dupsifter/releases).

Specifics about installation can be found in `docs/install.qmd` or on the
[documentation site](https://huishenlab.github.io/dupsifter/install.html).

## Examples

Examples of using `dupsifter` can be found in `docs/examples.qmd` or on the
[documentation site](https://huishenlab.github.io/dupsifter/examples.html). Test data can be found in the `example`
directory.

## Issues

Issues and bugs can be submitted to:
[https://github.com/huishenlab/dupsifter/issues](https://github.com/huishenlab/dupsifter/issues).

## Acknowledgments

  - This work is based on Gregory Faust's [samblaster](https://github.com/GregoryFaust/samblaster).
  - This work is supported by NIH/NCI R37CA230748.
