---
title: Usage
---

## Help

```default
Program: dupsifter
Version: 1.2.1
Contact: Jacob Morrison <jacob.morrison@vai.org>

dupsifter [options] <ref.fa> [in.bam]

Output options:
    -o, --output STR             name of output file [stdout]
    -O, --stats-output STR       name of file to write statistics to (see Note 3 for details)
Input options:
    -s, --single-end             run for single-end data
    -m, --add-mate-tags          add MC and MQ mate tags to mate reads
    -W, --wgs-only               process WGS reads instead of WGBS
    -l, --max-read-length INT    maximum read length for paired end duplicate-marking [10000]
    -b, --min-base-qual INT      minimum base quality [0]
    -B, --has-barcode            reads in file have barcodes (see Note 4 for details)
    -r, --remove-dups            toggle to remove marked duplicate
    -v, --verbose                print extra messages
    -h, --help                   this help
        --version                print version info and exit

Note 1, [in.bam] must be name sorted. If not provided, assume the input is stdin.
Note 2, assumes either ALL reads are paired-end (default) or single-end.
    If a singleton read is found in paired-end mode, the code will break nicely.
Note 3, defaults to dupsifter.stat if streaming or (-o basename).dupsifter.stat
    if the -o option is provided. If -o and -O are provided, then -O will be used.
Note 4, dupsifter first looks for a barcode in the CB SAM tag, then in the CR SAM tag, then
    tries to parse the read name. If the barcode is in the read name, it must be the last element
    and be separated by a ':' (i.e., @12345:678:9101112:1234_1:N:0:ACGTACGT). Any separators
    found in the barcode (e.g., '+' or '-') are treated as 'N's and the additional parts of the
    barcode are included up to a maximum length of 16 bases/characters. Barcodes are taken from
    read 1 in paired-end sequencing only.
```

## Option Descriptions

:::{.column-page-inset-right}
| Short Option |     Long Option     | Argument Type | Description                                                                        |
|:------------:|:--------------------|:-------------:|------------------------------------------------------------------------------------|
|    `-o`      | `--output`          |    string     | Name of output file (either .sam or .bam)                                          |
|    `-O`      | `--stats-output`    |    string     | Name of file to write statistics to (end with `.dupsifter.stat`)                   |
|    `-s`      | `--single-end`      |     none      | Run for single-end data (only do this if you know the data is SE)                  |
|    `-m`      | `--add-mate-tags`   |     none      | Add MC (mate CIGAR) and MQ (mate MAPQ) tags to mated reads                         |
|    `-W`      | `--wgs-only`        |     none      | Process WGS data instead of WGBS (see [Methodology](methodology.qmd) for differences in processing) |
|    `-l`      | `--max-read-length` |    integer    | Maximum read length (handles padding for reference genome windows)                 |
|    `-b`      | `--min-base-qual`   |    integer    | Minimum base quality (used in determining bisulfite strand if tags not provided)   |
|    `-B`      | `--has-barcode`     |     none      | Use when reads have cell barcodes and you want to mark duplicates accordingly      |
|    `-r`      | `--remove-dups`     |     none      | Remove reads that are flagged as duplicates                                        |
|    `-v`      | `--verbose`         |     none      | Print extra messages when running                                                  |
|    `-h`      | `--help`            |     none      | Print usage help message and exit                                                  |
|              | `--version`         |     none      | Print `dupsifter` version and exit                                                 |
:::
